### DISCUSSION TITLE
Trouble with Applying Glacier Mask in BlockwiseCoreg for DEM Coregistration

### DISCUSSION THREAD

**Discussion by cwywilson** [GlacioHack/xdem#621] [General]
Hi Everyone,

I'm trying to use blockwise_coreg = xdem.coreg.BlockwiseCoreg(xdem.coreg.NuthKaab(), subdivision=16) to coregister my photogrammetry-derived DEM (which contains a lot of noise) to a reference DEM with a glacier mask. 

However, after several attempts, it seems that BlockwiseCoreg doesn't properly apply the glacier mask during the fit and apply steps, and it's still performing coregistration on the glacier-covered areas.
![image](https://github.com/user-attachments/assets/d8f05769-99e8-4586-a544-a8a69b981015)
![image](https://github.com/user-attachments/assets/a6925cd4-65e8-4c57-986b-7b58bb763476)

The glacier mask works correctly when using xdem.coreg.NuthKaab() as a global coregistration.

![image](https://github.com/user-attachments/assets/fdd152a8-f77a-4f4f-a823-d3f4f3a79754)

Also, whenever I use BlockwiseCoreg, it takes up a lot of memory. Could this be a memory limitation affecting the process size? Is there an alternative way to handle this, such as using Dask to manage the memory load?

Could you help me understand how to fix this issue?

**Comment by rhugonnet** [comment]
Thanks @cwywilson for opening this! Might be linked to #604 that @quinto-clara is looking at right now.
Could you share a short code example, and print the content of your `blockwise_coreg._meta["step_meta"]` and `blockwise_coreg.stats()` (which has the number of inliers per block), so that we can reproduce this?

The blockwise functionality is undergoing a bit of a "revamp" as this code was on the older side, and less tested than the rest. It should become more robust in the future :wink:.
We're also working on implementing Dask support for this type of method (#525), but it's not there yet. 
For reducing current memory usage, maybe defining the global parameters linked to the parallelization of the `fit()` through `concurrent.futures.ThreadPoolExecutor()` would solve the issue... Any advice on this @erikmannerfelt?

**Reply by cwywilson** [reply]
Hi @rhugonnet,

Thanks for your quick reply! Below, I've included the `blockwise_coreg._meta["step_meta"]` and `blockwise_coreg.stats()`.

To provide some context, I'm working with historical imagery from 1959. The landscape has changed significantly since then, and the most reliable reference DEM I have is ArcticDEM. The photogrammetric DEM has a lot of noise due to these landscape changes. As a result, the stable terrain shows an uncertainty of over ±20m. I'm hoping to use blockwise_coreg to reduce this uncertainty in the stable areas (although the area is not stable). / 


[{'i': 0,
  'transform': Affine(2.0, 0.0, 569163.0867945289,
         0.0, -2.0, 8114714.940609187),
  'inlier_count': 4108609,
  'nmad': 8.432468481445312,
  'median': 0.21716309,
  'representative_x': 574752.0867945289,
  'representative_y': 8110515.940609187,
  'representative_val': 1250.3628,
  'subsample': 500000.0,
  'random_state': None,
  'subsample_final': 500000,
  'shift_x': 22.744000000000003,
  'shift_y': 2.9680000000000004,
  'shift_z': -9.973448775390807},
 {'i': 1,
  'transform': Affine(2.0, 0.0, 578687.0867945289,
         0.0, -2.0, 8114714.940609187),
  'inlier_count': 9992759,
  'nmad': 8.147965649414061,
  'median': 0.092041016,
  'representative_x': 583448.0867945289,
  'representative_y': 8111441.940609187,
  'representative_val': 1353.2506,
  'subsample': 500000.0,
  'random_state': None,
  'subsample_final': 500000,
  'shift_x': -5.169999999999999,
  'shift_y': -0.6579999999999999,
  'shift_z': -5.342702871398842},
 {'i': 2,
  'transform': Affine(2.0, 0.0, 588211.0867945289,
         0.0, -2.0, 8114714.940609187),
  'inlier_count': 2283698,
  'nmad': 11.351156249999999,
  'median': 0.52001953,
  'representative_x': 591862.0867945289,
  'representative_y': 8109845.940609187,
  'representative_val': 1193.8458,
  'subsample': 500000.0,
  'random_state': None,
  'subsample_final': 500000,
  'shift_x': 15.346,
  'shift_y': -8.268,
  'shift_z': -8.106338420287898},
 {'i': 4,
  'transform': Affine(2.0, 0.0, 569163.0867945289,
         0.0, -2.0, 8108168.940609187),
  'inlier_count': 9550951,
  'nmad': 11.394410815429687,
  'median': 0.34588623,
  'representative_x': 573924.0867945289,
  'representative_y': 8104897.940609187,
  'representative_val': 762.0505,
  'subsample': 500000.0,
  'random_state': None,
  'subsample_final': 500000,
  'shift_x': 23.346000000000004,
  'shift_y': -13.053999999999998,
  'shift_z': -18.668457841919064},
 {'i': 5,
  'transform': Affine(2.0, 0.0, 578687.0867945289,
         0.0, -2.0, 8108168.940609187),
  'inlier_count': 12605611,
  'nmad': 13.32358253173828,
  'median': 0.7525635,
  'representative_x': 583448.0867945289,
  'representative_y': 8104897.940609187,
  'representative_val': 957.9888,
  'subsample': 500000.0,
  'random_state': None,
  'subsample_final': 500000,
  'shift_x': -13.49,
  'shift_y': -22.408,
  'shift_z': -15.925935408325188},
 {'i': 6,
  'transform': Affine(2.0, 0.0, 588211.0867945289,
         0.0, -2.0, 8108168.940609187),
  'inlier_count': 10540240,
  'nmad': 10.336483776855468,
  'median': 0.4876709,
  'representative_x': 592972.0867945289,
  'representative_y': 8104935.940609187,
  'representative_val': 821.1603,
  'subsample': 500000.0,
  'random_state': None,
  'subsample_final': 500000,
  'shift_x': -1.8959999999999997,
  'shift_y': -13.970000000000002,
  'shift_z': -13.174010842285043},
 {'i': 7,
  'transform': Affine(2.0, 0.0, 597737.0867945289,
         0.0, -2.0, 8108168.940609187),
  'inlier_count': 811639,
  'nmad': 13.489271044921875,
  'median': 1.4297485,
  'representative_x': 600184.0867945289,
  'representative_y': 8102401.940609187,
  'representative_val': 699.3978,
  'subsample': 500000.0,
  'random_state': None,
  'subsample_final': 500000,
  'shift_x': -17.895999999999997,
  'shift_y': -29.474,
  'shift_z': -18.145688950195336},
 {'i': 8,
  'transform': Affine(2.0, 0.0, 569163.0867945289,
         0.0, -2.0, 8101624.940609187),
  'inlier_count': 1969810,
  'nmad': 14.008280584716797,
  'median': 0.70184326,
  'representative_x': 575956.0867945289,
  'representative_y': 8100059.940609187,
  'representative_val': 588.42773,
  'subsample': 500000.0,
  'random_state': None,
  'subsample_final': 500000,
  'shift_x': -4.565999999999999,
  'shift_y': -45.266,
  'shift_z': -29.31096953540043},
 {'i': 9,
  'transform': Affine(2.0, 0.0, 578687.0867945289,
         0.0, -2.0, 8101624.940609187),
  'inlier_count': 13168943,
  'nmad': 7.010768737792969,
  'median': 0.022766113,
  'representative_x': 583448.0867945289,
  'representative_y': 8098351.940609187,
  'representative_val': 450.13065,
  'subsample': 500000.0,
  'random_state': None,
  'subsample_final': 500000,
  'shift_x': 1.068,
  'shift_y': 1.83,
  'shift_z': -4.020495444030644},
 {'i': 10,
  'transform': Affine(2.0, 0.0, 588211.0867945289,
         0.0, -2.0, 8101624.940609187),
  'inlier_count': 13435322,
  'nmad': 29.330214990234374,
  'median': 2.8953552,
  'representative_x': 592974.0867945289,
  'representative_y': 8098351.940609187,
  'representative_val': 270.07623,
  'subsample': 500000.0,
  'random_state': None,
  'subsample_final': 500000,
  'shift_x': -50.956,
  'shift_y': -34.82200000000001,
  'shift_z': -33.3873942005921},
 {'i': 11,
  'transform': Affine(2.0, 0.0, 597737.0867945289,
         0.0, -2.0, 8101624.940609187),
  'inlier_count': 6851663,
  'nmad': 24.65492131347656,
  'median': 2.005005,
  'representative_x': 602498.0867945289,
  'representative_y': 8098351.940609187,
  'representative_val': 791.7659,
  'subsample': 500000.0,
  'random_state': None,
  'subsample_final': 500000,
  'shift_x': -3.2659999999999996,
  'shift_y': 14.687999999999999,
  'shift_z': -5.4853220161137415},
 {'i': 13,
  'transform': Affine(2.0, 0.0, 578687.0867945289,
         0.0, -2.0, 8095078.940609187),
  'inlier_count': 5738301,
  'nmad': 2.0542298950195312,
  'median': 0.01977539,
  'representative_x': 583448.0867945289,
  'representative_y': 8091805.940609187,
  'representative_val': 402.40735,
  'subsample': 500000.0,
  'random_state': None,
  'subsample_final': 500000,
  'shift_x': -4.244,
  'shift_y': 2.0759999999999996,
  'shift_z': 2.3177275480956894},
 {'i': 14,
  'transform': Affine(2.0, 0.0, 588211.0867945289,
         0.0, -2.0, 8095078.940609187),
  'inlier_count': 7157844,
  'nmad': 13.596723122406004,
  'median': 0.19565582,
  'representative_x': 592974.0867945289,
  'representative_y': 8091805.940609187,
  'representative_val': 56.29129,
  'subsample': 500000.0,
  'random_state': None,
  'subsample_final': 500000,
  'shift_x': 0.862,
  'shift_y': -0.726,
  'shift_z': -1.360232364692635},
 {'i': 15,
  'transform': Affine(2.0, 0.0, 597737.0867945289,
         0.0, -2.0, 8095078.940609187),
  'inlier_count': 348044,
  'nmad': 17.297030163574217,
  'median': 0.33276367,
  'representative_x': 600622.0867945289,
  'representative_y': 8094869.940609187,
  'representative_val': 849.48834,
  'subsample': 500000.0,
  'random_state': None,
  'subsample_final': 348044,
  'shift_x': -35.03999999999999,
  'shift_y': -7.7540000000000004,
  'shift_z': -10.715174056396563}]



center_x	center_y	center_z	x_off	y_off	z_off	inlier_count	nmad	median
chunk									
0	574752.086795	8.110516e+06	1250.362793	22.744	2.968	-9.973449	4108609	8.432468	0.217163
1	583448.086795	8.111442e+06	1353.250610	-5.170	-0.658	-5.342703	9992759	8.147966	0.092041
2	591862.086795	8.109846e+06	1193.845825	15.346	-8.268	-8.106338	2283698	11.351156	0.520020
3	583448.086795	8.104898e+06	957.988770	-13.490	-22.408	-15.925935	9550951	11.394411	0.345886
4	592972.086795	8.104936e+06	821.160278	-1.896	-13.970	-13.174011	12605611	13.323583	0.752563
5	600184.086795	8.102402e+06	699.397827	-17.896	-29.474	-18.145689	10540240	10.336484	0.487671
6	575956.086795	8.100060e+06	588.427734	-4.566	-45.266	-29.310970	811639	13.489271	1.429749
7	583448.086795	8.098352e+06	450.130646	1.068	1.830	-4.020495	1969810	14.008281	0.701843
8	592974.086795	8.098352e+06	270.076233	-50.956	-34.822	-33.387394	13168943	7.010769	0.022766
9	602498.086795	8.098352e+06	791.765930	-3.266	14.688	-5.485322	13435322	29.330215	2.895355
10	583448.086795	8.091806e+06	402.407349	-4.244	2.076	2.317728	6851663	24.654921	2.005005
11	600622.086795	8.094870e+06	849.488342	-35.040	-7.754	-10.715174	5738301	2.054230	0.019775

**Reply by rhugonnet** [reply]
Thanks @cwywilson. 
Everything is looking normal, although it's a lot of inlier pixels per block. Can you compare the sum of those `inlier_count` for all blocks, with the number of inliers in your original mask for the entire DEM? (`np.count_nonzero(inlier_mask)`).
This would be the definitive check that the issue comes from the `inlier_mask`.

Otherwise, could the large vertical shifts be due to the `inlier_mask` containing large amounts of unmasked glacierized terrain (if the DEM is old but not the outlines) that has retreated since then?

**Reply by cwywilson** [reply]
Hi @rhugonnet,

Apologies for the delay in getting back to you. The blockwise code takes a significant amount of time to run, and due to the large memory requirements, it often crashes at the end.

However, I noticed an issue: the` blockwise_coreg.stats().inlier_count` is higher than` np.count_nonzero(inlier_mask)`, and I’m unsure why this discrepancy occurs. I’d appreciate any insights you might have. Here’s an overview of the code I’m using:

I have refdem, demtobealign, and glacier_outline defined as:
`glacier_outlines = gu.Vector("OUTLINE.shp")`
I also have a slope mask:
```
slope = xdem.terrain.slope(reference_dem)

# Create masks for slopes between 0-5 degrees and slopes over 30 degrees
slope_mask_low = (slope.data <= 5)  # Mask for slopes between 0-5 degrees
slope_mask_high = (slope.data >= 30)  # Mask for slopes over 30 degrees

# Combine the masks (0-5 degrees or >= 30 degrees)
slope_combined_mask = slope_mask_low | slope_mask_high
```
The inlier mask is defined as:
```
glacier_mask = glacier_outlines.create_mask(reference_dem)
combined_mask = slope_combined_mask | glacier_mask.data
inlier_mask = ~combined_mask
```

When I count the inliers using `np.count_nonzero(inlier_mask)`, I get 84,965,169. However, after running `blockwise_coreg`, the reported `inlier_coun`t is 105,454,194.

If needed, I can upload both DEMs and the notebook for you to review.

Thank you for your help!

Wilson

**Reply by rhugonnet** [reply]
Thanks @cwywilson,

EDIT: After more investigation, I found the bug! Indeed the mask was ignored :smile: 
The `inlier_mask` variable is used twice for two different purposes in this function, overwriting the previous one: https://github.com/GlacioHack/xdem/blob/5bde36e985ff1179a0eb4c465822f5461adc72f0/xdem/coreg/base.py#L2964
It's a 2-line fix!

For memory usage: Until we finalize Dask support of this, the best would likely be for you to manually split your DEM into tiles (with some overlap), and save those as smaller files. It will then be extremely fast to co-register those with a simple Nuth and Kääb, and later merge them back together (for example with: https://geoutils.readthedocs.io/en/stable/gen_modules/geoutils.raster.merge_rasters.html#geoutils.raster.merge_rasters) :wink:.

**Reply by rhugonnet** [reply]
Opened #624 and resolved in #625. 
It'll take a bit of time (probably at least a month) until we can publish a new release, because we also have to fix #584 for `BlockwiseCoreg` (this was not an issue for you running the code on the published xDEM versions, it's only a recent development issue on the latest GitHub main due to other changes).

**Reply by rhugonnet** [reply]
Thanks again for reporting this @cwywilson, good catch!
As I mentioned above, the `BlockwiseCoreg` was an old functionality with not so many tests. We are improving it to be more modular, robust (this will come soon), and computationally efficient (this will take a while longer but is on the way too).

**Reply by cwywilson** [reply]
I’m glad we could identify the issues—thank you for your efforts. XDEM is indeed a powerful tool and has great potential, as demonstrated in this study: https://tc.copernicus.org/articles/18/3195/2024/, as well as in historical image reconstruction. I would love to contribute to the XDEM project in the future. However, as I’m in the final year of my PhD, I need to focus on completing my current work before taking on new tasks.
